core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/util/CanonicalRequestV2.java

```diff
@@ -0,0 +1,329 @@
+
+                encodedValues.add(signatureFormattedEncodedValue);
+            }
+            Collections.sort(encodedValues);
+            sorted.put(encodedParamName, encodedValues);
+
+        });
+        return sorted;
+    }
+
+    /**
+     * Get the string representing query string parameters.
+     * Parameters are URL-encoded and separated by an ampersand.
+     * <p>
+     * Reserved characters are percent-encoded, names and values are encoded separately
+     * and empty parameters have an equals-sign appended before encoding.
+     * <p>
+     * After encoding, parameters are sorted alphanetically by key name.
+     * <p>
+     * If no query string is given, an empty string ("") is returned.
+     */
+    public static String getCanonicalQueryString(SortedMap<String, List<String>> canonicalParams) {
+        StringBuilder stringBuilder = new StringBuilder(512);
+        SdkHttpUtils.flattenQueryParameters(stringBuilder, canonicalParams);
+
+        return stringBuilder.toString();
+    }
+
+    /**
+     * Get the list of headers that are to be signed.
+     */
+    public static List<Pair<String, List<String>>> getCanonicalHeaders(Map<String, List<String>> headers) {
+        List<Pair<String, List<String>>> result = new ArrayList<>(headers.size());
+
+        // headers retrieved from the request are already sorted case-insensitively
+        headers.forEach((key, value) -> {
+            String lowerCaseHeader = lowerCase(key);
+            if (!HEADERS_TO_IGNORE_IN_LOWER_CASE.contains(lowerCaseHeader)) {
+                result.add(Pair.of(lowerCaseHeader, value));
+            }
+        });
+
+        result.sort(Comparator.comparing(Pair::left));
+
+        return result;
+    }
+
+    /**
+     * Get the string representing the headers that will be signed and their values.
+     * Headers names are lower-case, sorted in alphabetical order, and are followed by a colon.
+     * <p>
+     * Values are trimmed of any leading/trailing spaces, sequential spaces are converted to single
+     * space, and multiple values are comma separated.
+     * <p>
+     * Each header-value pair is separated by a newline.
+     */
+    public static String getCanonicalHeadersString(List<Pair<String, List<String>>> canonicalHeaders) {
+        StringBuilder result = new StringBuilder(512);
+        canonicalHeaders.forEach(header -> {
+            result.append(header.left());
+            result.append(":");
+            for (String headerValue : header.right()) {
+                addAndTrim(result, headerValue);
+                result.append(",");
+            }
+            result.setLength(result.length() - 1);
+            result.append("\n");
+        });
+        return result.toString();
+    }
+
+    /**
+     * Get the string representing which headers are part of the signing process.
+     * Header names are lower-case, sorted in alphabetical order, and separated by a semicolon.
```

Comment:
the lower casing and alphabetical ordering is not part of this method. maybe that part of the comment is meant for getCanonicalHeaders